torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -330,27 +388,47 @@ def _del_promise_file(promise_filename, filename):
             )
 
     def __iter__(self):
-        old_promise_filename = None
-        old_filename = None
-        first_entry = True
+        # old_promise_filename = None
+        # old_filename = None
+        # first_entry = True
+        last_record_uuid = None
         # TODO(VitalyFedyunin): Limit buffer size here. It is only contains file names from archive,
-        # but better be save than sorry.
-        buffer: List[Any] = []
+        # but better be safe than sorry.
+        # buffer: List[Any] = []
+        one_to_many_detected = False
+        one_to_one_detected = False
+
+        def fulfill_old_promises(buffer, last_record_uuid, first_filepath_fn):
+            for old_rec_uuid, old_rec in buffer:
+                original_file_name = first_filepath_fn(old_rec)
+                old_promise_filename = _find_promise_file(original_file_name)
+                self._del_promise_file(old_promise_filename, original_file_name)
+                if old_rec_uuid == last_record_uuid:
+                    break
+
         for filename in self.source_datapipe:
-            promise_filename = _find_promise_file(filename)
-            if not first_entry:
-                buffer.append(old_filename)
-                if old_promise_filename != promise_filename:
-                    self._del_promise_file(old_promise_filename, old_filename)
-                    yield from buffer
-                    buffer = []
-            old_promise_filename = promise_filename
-            old_filename = filename
-            first_entry = False
-        if not first_entry:
-            buffer.append(old_filename)
-            self._del_promise_file(old_promise_filename, old_filename)
-            yield from buffer
+            rec_uuid, record = self.memory_cell_dp.get_last()
+            original_file_name = self.first_filepath_fn(record)
+            if original_file_name != filename:
+                one_to_many_detected = True
+                if one_to_one_detected:
+                    raise Exception("DIsovered different keys when one-to-one mode previously assumed")
+                # We are dealing with one-to-many situation now
+                with open(_get_list_filename(original_file_name), "a") as fh:
+                    fh.write(f"{filename}\n")
+            else:
+                one_to_one_detected = True
+                if one_to_many_detected:
+                    # Keys should be always the same (1-1 situation) or always different (1-many) sutuation
+                    raise Exception("first key somehow equal to secondary key")
+            print(f"{filename} generateed from rec {rec_uuid} {record}")
+            if rec_uuid != last_record_uuid:
+                fulfill_old_promises(self.memory_cell_dp.get_buffer()[1:], last_record_uuid, self.first_filepath_fn)
+                last_record_uuid = rec_uuid
```

Comment:
It is already happening at line 412

Reply:
I mean we need to set `one_to_many_detected` and `one_to_one_detected` to `False` after `last_record_uuid` is updated (line 427)

Reply:
I think not, but I think I will reconsider entire detection as it will fail out if the first one-to-many will be one-to-one. All code will misjudge pipeline as 1:1 in this case.

Reply:
(anyway I'm working on more tests now, will update code shortly)

Reply:
Done, should work fine now.